barretenberg/cpp/src/barretenberg/honk/composer/ultra_composer.test.cpp

```diff
@@ -884,6 +868,66 @@ TEST_F(UltraHonkComposerTests, ram)
     prove_and_verify(circuit_builder, composer, /*expected_result=*/true);
 }
 
+TEST_F(UltraHonkComposerTests, range_checks_on_duplicates)
+{
+    auto circuit_builder = proof_system::UltraCircuitBuilder();
+
+    uint32_t a = circuit_builder.add_variable(100);
+    uint32_t b = circuit_builder.add_variable(100);
+    uint32_t c = circuit_builder.add_variable(100);
+    uint32_t d = circuit_builder.add_variable(100);
+
+    circuit_builder.assert_equal(a, b);
+    circuit_builder.assert_equal(a, c);
+    circuit_builder.assert_equal(a, d);
+
+    circuit_builder.create_new_range_constraint(a, 1000);
+    circuit_builder.create_new_range_constraint(b, 1001);
+    circuit_builder.create_new_range_constraint(c, 999);
+    circuit_builder.create_new_range_constraint(d, 1000);
+
+    circuit_builder.create_big_add_gate(
+        {
+            a,
+            b,
+            c,
+            d,
+            0,
+            0,
+            0,
+            0,
+            0,
+        },
+        false);
+
+    auto composer = UltraComposer();
+    prove_and_verify(circuit_builder, composer, /*expected_result=*/true);
+}
+
+// Ensure copy constraints added on variables smaller than 2^14, which have been previously
+// range constrained, do not break the set equivalence checks because of indices mismatch.
+// 2^14 is DEFAULT_PLOOKUP_RANGE_BITNUM i.e. the maximum size before a variable gets sliced
+// before range constraints are applied to it.
+TEST_F(UltraHonkComposerTests, range_constraint_small_variable)
+{
+    auto circuit_builder = proof_system::UltraCircuitBuilder();
+
+    uint16_t mask = (1 << 8) - 1;
+    int a = engine.get_random_uint16() & mask;
+    uint32_t a_idx = circuit_builder.add_variable(fr(a));
+    uint32_t b_idx = circuit_builder.add_variable(fr(a));
+    ASSERT_NE(a_idx, b_idx);
+    uint32_t c_idx = circuit_builder.add_variable(fr(a));
+    ASSERT_NE(c_idx, b_idx);
+    circuit_builder.create_range_constraint(b_idx, 8, "bad range");
```

Comment:
is this 8 arbitrary or does it have meaning related to range constraints?

Reply:
I believe it's arbitrary other than 8 < 14 per the comment. I don't have full context here but I think the point is simply to check that range constraining a value (small enough to require only 1 range constraint) prior to copy constraining it to another value yields a valid proof. Apparently previously there was a bug that caused such an operation to fail.